File: #1235.maximum-profit-in-job-scheduling.cpp
class Solution {
public:
    int jobScheduling(vector<int>& startTime, vector<int>& endTime, vector<int>& profit) {

        int size = startTime.size();
        vector<array<int, 3>> ar;
        for (int i = 0; i < size; ++i) {
            ar.push_back({startTime[i], endTime[i], profit[i]});
        }

        std::sort(ar.begin(), ar.end());
        
        map<int,int> m1, m2;
        int result = -1;
        for (int i = size - 1; i >= 0; --i) {
            auto it = m1.lower_bound(ar[i][1]);
            if (it != m1.end()) {
                m2[ar[i][0]] = std::max({result, (*it).second + ar[i][2]});
            } else {
                m2[ar[i][0]] = std::max({ar[i][2], result, m2[ar[i][0]]});
            }
            
            m1[ar[i][0]] = std::max(m1[ar[i][0]], m2[ar[i][0]]);
            result = std::max(result, m2[ar[i][0]]);
        }
        return result;
    }
};

File: #1235.maximum-profit-in-job-scheduling.js
/**
 * @param {number[]} startTime
 * @param {number[]} endTime
 * @param {number[]} profit
 * @return {number}
 */
var jobScheduling = function(startTime, endTime, profit) {
    let jobs = [];
    let length = startTime.length;
    for (let i = 0; i < length; i++) {
        jobs.push({s: startTime[i], e: endTime[i], p: profit[i]});
    }

    jobs.sort(function(x,y) {return x.e - y.e});
    let dp = new Array(length);
    dp[0] = jobs[0].p;

    for (let i = 1; i < length; i++) {
        let profit = jobs[i].p;
        let task = -1;
        for (let j = i-1; j >= 0; j--) {
            if (jobs[j].e <= jobs[i].s) {
                task = j;
                break;
            }
        }
        if (task != -1) {
            profit += dp[task];
        }
        dp[i] = Math.max(profit, dp[i - 1]);
    }
    return dp[length - 1];
};
